include/boost/geometry/algorithms/detail/closest_points/segment_to_segment.hpp

```diff
@@ -0,0 +1,154 @@
+{
+
+
+
+// compute segment-segment closest-points
+class segment_to_segment
+{
+public:
+    
+    template <typename Segment1, typename Segment2, typename OutputSegment, typename Strategies>
+    static inline void apply(Segment1 const& segment1, Segment2 const& segment2,
+                             OutputSegment& shortest_seg,
+                             Strategies const& strategies)
+    {
+        using intersection_return_type = segment_intersection_points
+            <
+                typename point_type<Segment1>::type
+            >;
+
+        using intersection_policy = policies::relate::segments_intersection_points
+            <
+                intersection_return_type
+            >;
+
+        detail::segment_as_subrange<Segment1> sub_range1(segment1);
+        detail::segment_as_subrange<Segment2> sub_range2(segment2);
+        auto is = strategies.relate().apply(sub_range1, sub_range2,
+                                            intersection_policy());
+        if (is.count > 0)
+        {
+            set_segment_from_points::apply(is.intersections[0], 
+                                           is.intersections[0], 
+                                           shortest_seg);
+            return;
+        }
+        
+        typename point_type<Segment1>::type p[2];
+        detail::assign_point_from_index<0>(segment1, p[0]);
+        detail::assign_point_from_index<1>(segment1, p[1]);
+
+        typename point_type<Segment2>::type q[2];
+        detail::assign_point_from_index<0>(segment2, q[0]);
+        detail::assign_point_from_index<1>(segment2, q[1]);
+
+        auto cp0 = strategies.closest_points(q[0], segment1).apply(q[0], p[0], p[1]);
+        auto cp1 = strategies.closest_points(q[1], segment1).apply(q[1], p[0], p[1]);
+        auto cp2 = strategies.closest_points(p[0], segment2).apply(p[0], q[0], q[1]);
+        auto cp3 = strategies.closest_points(p[1], segment2).apply(p[1], q[0], q[1]);
+
+        closest_points::creturn_t<Segment1, Segment2, Strategies> d[4];
+        
+        //TODO: this could be simplified by getting distance strategy in a unique way
+        d[0] = strategy::distance::services::get_comparable
+            <
+                decltype(strategies.distance(cp0, q[0]))
+            >::apply(strategies.distance(cp0, q[0])).apply(cp0, q[0]);
+        d[1] = strategy::distance::services::get_comparable
+            <
+                decltype(strategies.distance(cp1, q[1]))
+            >::apply(strategies.distance(cp1, q[1])).apply(cp1, q[1]);
+        d[2] = strategy::distance::services::get_comparable
+            <
+                decltype(strategies.distance(p[0], cp2))
+            >::apply(strategies.distance(p[0], cp2)).apply(p[0], cp2);
+        d[3] = strategy::distance::services::get_comparable
+            <
+                decltype(strategies.distance(p[1], cp3))
+            >::apply(strategies.distance(p[1], cp3)).apply(p[1], cp3);
```

Comment:
All of these calculate point to point distances right?

If that's the case you only need to create one strategy, e.g. using the first pair of points. But there is actually no need to pass the actual points. You could pass `detail::dummy_point`s [1]. This way you could create the strategy somewhere above where the points are unavailable. Furthermore instead of `get_comparable` you could use `strategies::distance::detail::make_comparable()` [2] to make umbrella strategy returning comparable legacy strategies from `distance()` member function. Though this would be an additional dependency so you might want to choose to stick with `get_comparable` but using it only once. Anyway, this is how this code could be reimplemented:
```
auto const cds = strategies::distance::detail::make_comparable(strategies)
                     .distance(detail::dummy_point(), detail::dummy_point());
d[0] = cds.apply(cp0, q[0]);
d[1] = cds.apply(cp1, q[1]);
// etc.
```

[1] https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/algorithms/detail/dummy_geometries.hpp#L26
[2] https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/strategies/distance/comparable.hpp#L51